qiskit/dagcircuit/collect_blocks.py

```diff
@@ -86,22 +87,48 @@ def _op_nodes(self):
             return self.dag.get_nodes()
 
     def _direct_preds(self, node):
-        """Returns direct predecessors of a node."""
+        """Returns direct predecessors of a node. This function takes into account the
+        direction of collecting blocks, that is node's predecessors when collecting
+        backwards are the direct successors of a node in the DAG.
+        """
         if not self.is_dag_dependency:
-            return [pred for pred in self.dag.predecessors(node) if isinstance(pred, DAGOpNode)]
+            if self._collect_from_back:
+                return [pred for pred in self.dag.successors(node) if isinstance(pred, DAGOpNode)]
```

Comment:
I find this a little confusing because we're really reversing the edges in this case, but the `preds` method is calling the `successors` method. I'm wondering if instead we can use `DAGCircuit.reverse_ops()` or something to actually reverse the edges and then call `predecessors`?

Reply:
Thanks! I agree that `reverse_ops()` is weird, and should be used with care. In fact, my very original suggestion to Petar was to call `reverse_ops()` on the circuit, collect-and-collapse blocks as usual, and call `reverse_ops()` on the resulting circuit; however, this suggestion was not correct. As a tiny example, suppose that we have gates `H(0), CX(0, 1), CX(1, 2), S(2)` and we want to replace the two CX gates by a linear function but starting from the back of the circuit. Reversing, we get `S(2), CX(1, 2), CX(0, 1), H(0)`, collecting-and-collapsing blocks as usual would give us `S(2), L, H(0)` where `L` is the linear function for `CX(1, 2), CX(0, 1)`, reversing again gives us `H(0), L, S(2)`. However, this is incorrect, as we want the linear function for `CX(0, 1) CX(1, 2)` and not the other way around. I do agree that reversing the edges of the DAG, collecting blocks, reversing collected blocks, and then collapsing also seems correct.